File: typescript/js/word_pattern.js
"use strict";
var WordPattern = /** @class */ (function () {
    function WordPattern(pattern, str) {
        this.pattern = pattern;
        this.str = str;
        this.pattern = pattern;
        this.str = str;
    }
    WordPattern.prototype.is_match = function () {
        var splitStr = this.str.split(" ");
        var result = {};
        if (this.pattern.length !== splitStr.length)
            return false;
        for (var i = 0; i < this.pattern.length; i++) {
            if (!Object.keys(result).includes(this.pattern[i])) {
                if (Object.values(result).includes(splitStr[i])) {
                    return false;
                }
                result[this.pattern[i]] = splitStr[i];
                console.log(result);
            }
            else if (result[this.pattern[i]] !== splitStr[i]) {
                return false;
            }
        }
        return true;
    };
    return WordPattern;
}());
var patternObj = new WordPattern("abbb", "dog cat cat dog");
console.log(patternObj.is_match());

File: python/word_pattern.py
class WordPattern:
    def __init__(self, pattern: str, string:str):
        self.pattern = pattern
        self.string = string
    
    # Both Time and space efficient
    def is_match(self):
        split_str = self.string.split(' ')
        if len(self.pattern) != len(split_str):
            return False
        ch_to_word = {}
        word_to_char = {}

        for i in range(len(self.pattern)):
            if self.pattern[i] in ch_to_word and ch_to_word[self.pattern[i]] != split_str[i]:
                return False
            if split_str[i] in word_to_char and word_to_char[split_str[i]] != self.pattern[i]:
                return False
            
            ch_to_word[self.pattern[i]] = split_str[i]
            word_to_char[split_str[i]] = self.pattern[i]
        
        return True

    # General solution
    def is_match1(self):
        result = {}
        split_str = self.string.split()
        if len(self.pattern) != len(split_str):
            return False
        for i in range(len(self.pattern)):
            if self.pattern[i] not in result:
                if split_str[i] in result.values():
                    return False
                result[self.pattern[i]] = split_str[i]
            elif split_str[i] != result[self.pattern[i]]:
                return False
        print(result)
        return True


word_obj = WordPattern('abba', 'dog cat cat fish')

print(word_obj.is_match())